src/main/java/tdl/s3/upload/FileUploadingService.java

```diff
@@ -1,76 +1,53 @@
 package tdl.s3.upload;
 
 import com.amazonaws.services.s3.AmazonS3;
-import com.amazonaws.services.s3.model.ListMultipartUploadsRequest;
 import com.amazonaws.services.s3.model.MultipartUpload;
-import com.amazonaws.services.s3.model.MultipartUploadListing;
 
 import java.io.File;
-import java.util.Collection;
-import java.util.List;
-import java.util.stream.Collectors;
-import java.util.stream.Stream;
+import tdl.s3.helpers.ExistingMultipartUploadFinder;
+import tdl.s3.sync.SyncProgressListener;
 
 public class FileUploadingService {
+
     private final AmazonS3 amazonS3;
+
     private final String bucket;
-    private String prefix;
+
+    private final String prefix;
+
+    private SyncProgressListener listener;
 
     public FileUploadingService(AmazonS3 amazonS3, String bucket, String prefix) {
         this.amazonS3 = amazonS3;
         this.bucket = bucket;
-	    this.prefix = prefix;
+        this.prefix = prefix;
+    }
+
+    public void setListener(SyncProgressListener listener) {
+        this.listener = listener;
     }
 
     public void upload(File file) {
         RemoteFile remoteFile = new RemoteFile(bucket, prefix, file.getName(), amazonS3);
         upload(file, remoteFile);
     }
-    
+
     public void upload(File file, String remoteName) {
         RemoteFile remoteFile = new RemoteFile(bucket, prefix, remoteName, amazonS3);
         upload(file, remoteFile);
     }
 
     private void upload(File file, RemoteFile remoteFile) {
-        FileUploader fileUploader = bringFileUploader(remoteFile);
+        FileUploader fileUploader = createFileUploader(remoteFile);
         fileUploader.upload(file, remoteFile);
     }
 
-    private FileUploader bringFileUploader(RemoteFile remoteFile) {
-        List<MultipartUpload> alreadyStartedUploads = getMultipartUploads(amazonS3, remoteFile.getBucket(), remoteFile.getPrefix());
-        MultipartUpload multipartUpload = alreadyStartedUploads.stream()
-                .filter(upload -> upload.getKey().equals(remoteFile.getFullPath()))
-                .findAny()
-                .orElse(null);
+    private FileUploader createFileUploader(RemoteFile remoteFile) {
+        ExistingMultipartUploadFinder finder = new ExistingMultipartUploadFinder(amazonS3, remoteFile.getBucket(), remoteFile.getPrefix());
+        MultipartUpload multipartUpload = finder.findOrNull(remoteFile);
+
         UploadingStrategy strategy = new MultiPartUploadFileUploadingStrategy(multipartUpload);
+        strategy.setListener(listener);
```

Comment:
To make sure we always have the right value for this listener, I would make it a mandatory constructor parameter.